Python/syntax_test_python_strings.py

```diff
@@ -549,6 +549,45 @@
 bar = "}}" # Comment
 #      ^^ constant.character.escape
 
+# The following section contains unusual and legal or illegal format placeholders.
+# We don't actually want to match the syntax 100% of the time,
+# since we never know for sure if the string is used as a format string,
+# so some of these matches are incorrect because of implementation details.
+
+# Not format specs
+"{:{ }"  # unclosed
+# ^ - constant.other.placeholder
+'{{foo!r:4.2}'  # escaped opening
+# ^ - constant.other.placeholder
+'{{foo!r:4.2}}'  # escaped opening and closing
+# ^ - constant.other.placeholder
+'{foo!a:ran{dom}'  # unclosed
+# ^ - constant.other.placeholder
+'{foo!a:ran{dom}'  # unclosed
+# ^ - constant.other.placeholder
+
+# Invalid field names
+'{foo{d}}'
+# ^ - constant.other.placeholder
+'{foo.!a:d}'
+# ^ constant.other.placeholder
+"{:{ {}}"  # issue 2232
+# ^ - constant.other.placeholder
+
+# Syntactically correct, but hardly come up in real code
+"{:{ ()}}".format(0, **{" ()": "d"}) == '0'
+# ^ constant.other.placeholder
+'{foo/bar}'.format(**{"foo/bar": 1}) == '1'
+# ^ - constant.other.placeholder
+
+# Legal but non-standard format spec
+'{foo:{{w}}.{{p}}}}'  # illegal double {{ in format spec
```

Comment:
This line confuses me. Above it says "Legal but non-standard", whereas this line says "illegal".

Reply:
The syntax is legal and translates to `format(foo, "{w}.{p}")`. The comment refers to how a format spec of `{w}.{p}` is unusual and must be interpreted by a custom `__format__` implementation, because this is unheard of in the entire stdlib. I'll reword it to be clearer.

Reply:
Actually, the spec is illegal because `}` (or rather `}}` in the escaped form) doesn't work. This doesn't seem to be documented anywhere, but the following snippet can be used to test behavior:
```py
class Foo:
    def __format__(self, spec):
        return spec.format(w=1, p=123)
fmt = '{foo:{{w}}.{{p}}}}'
print(fmt.format(foo=Foo()))
```
```
Traceback (most recent call last):
  File "/tmp/format-spec.py", line 5, in <module>
    print(fmt.format(foo=Foo()))
ValueError: Single '}' encountered in format string
```

Reply:
Here's an example implementation showing how it works:
```py
class Foo:
    def __format__(self, spec):
        return spec.format(w=1, p=123)
fmt = '{foo:{{w}}.{{p}}}'
print(fmt.format(foo=Foo()))
```
```
1.123
```